### PR TITLE
Fix the derivative of `acosh` for complex numbers.

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80841

For a summary of what was going on, see https://github.com/pytorch/pytorch/issues/50692#issuecomment-1173922441

Fixes https://github.com/pytorch/pytorch/issues/50692

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -277,7 +277,7 @@
   output_differentiability: [False]
 
 - name: acosh(Tensor self) -> Tensor
-  self: grad * (self.pow(2) - 1).rsqrt().conj()
+  self: "self.is_complex() ? grad * ((self + 1).rsqrt() * (self - 1).rsqrt()).conj() : grad * (self * self - 1).rsqrt()"
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:280]: The two are mathematically the same at a glance? Why can't we keep only one?

Comment (Lezcano) [reply tools/autograd/derivatives.yaml:280]: The first one is correct for both real and complex numbers. The second one is a simplification for real numbers.

Comment (albanD) [reply tools/autograd/derivatives.yaml:280]: Can you leave a comment there: save one addition by re-ordering the terms for the real case.

**Review by albanD — APPROVED** [review]
I think a small comment to explain the duplication would be nice.
LGTM otherwise!

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (Lezcano) [reply tools/autograd/derivatives.yaml:281]: Ugh, of course I wrote "compelx"... lol

**Comment by anjali411** [comment]
Similarly looks like atanh derivative is also wrong. It should be: `grad * 1 / ((1 - self) * (1 + self)).conj()`

**Comment by Lezcano** [comment]
I think the derivative for the `atanh` is right.

**Comment by Lezcano** [comment]
Re the edit. Those notes are for real numbers. The formula for the solution of a quadratic equation just works for real coefficients. For complex coefficients, you need to "complete the square" (which is the way to prove the usual formula for the quadratic equation really) and be careful when taking square roots and logarithms, as these are multivalued functions.

If you to simply get a formula, ask what's `asin'(z)` to wolfram alpha.

**Comment by Lezcano** [comment]
fwiw, cppreference is also careful of defining things the right way in https://en.cppreference.com/w/cpp/numeric/complex/acosh.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:281]: This can be added to any other PR :D

**Pull request closed without merging**